### PR TITLE
Rewrite ckeditor5-undo to TypeScript

### LINKED ISSUE #12612: [TS MVP] Rewrite ckeditor5-undo

### PR DESCRIPTION
**Pull request by arkflpc**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (undo): Rewrites ckeditor5-undo to TypeScript. Closes #12612.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-engine/src/controller/datacontroller.ts [hunk 1]
```diff
@@ -639,3 +639,21 @@ function _getMarkersRelativeToElement( element: ModelElement ): Map<string, Mode
...
+	name: 'set';
+	args: [ Parameters<DataController[ 'set' ]> ];
+	return: ReturnType<DataController[ 'set' ]>;
+};
+
+export type DataControllerGetEvent = {
+	name: 'get';
+	args: [ Parameters<DataController[ 'get' ]> ];
+	return: ReturnType<DataController[ 'get' ]>;
+};
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/controller/datacontroller.ts:659]: Why you are not using `DecoratedMethodEvent<...>` that we already have?

Comment (arkflpc) [reply packages/ckeditor5-engine/src/controller/datacontroller.ts:659]: Because it's not Observable, it just pretends to be.

Comment (niegowski) [reply packages/ckeditor5-engine/src/controller/datacontroller.ts:659]: Its decorated method, not observable property ;)

File: packages/ckeditor5-undo/src/basecommand.ts [hunk 2]
```diff
@@ -17,7 +22,14 @@ import { transformSets } from '@ckeditor/ckeditor5-engine/src/model/operation/tr
  * @extends module:core/command~Command
  */
 export default class BaseCommand extends Command {
```
Comment (arkflpc) [inline packages/ckeditor5-undo/src/basecommand.ts:24]:
```suggestion
export default abstract class BaseCommand extends Command {
```

**Pull request merged**